src/main/java/com/google/sps/servlets/PhotoHandlerServlet.java

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+import com.google.gson.Gson;
+
+import com.google.sps.GoogleShoppingQuerier;
+import com.google.sps.ShoppingQuerierConnectionException;
+import com.google.sps.data.Product;
+import com.google.sps.data.ShoppingQueryInput;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+
+/**
+ * When the user submits the form for image uploading, Blobstore processes the file upload and forwards
+ * the request to this servlet, which returns the product shopping results for the respective photo,
+ * in JSON format.
+ */
+@WebServlet("/handle-photo")
+public class PhotoHandlerServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey blobKey = getBlobKey(request, "photo");
+
+    // Send an error if the user did not upload a file.
+    if (blobKey == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Client must upload an image file.");
+      return;
+    }
```

Comment:
`getBlobKey` returns null if the user did not upload a photo, therefore it is okay to have that if statement there. The error cannot be thrown from that method because the error is set from the response, which we handle in this doPost() method, not in the other one.

Reply:
Yeh so you can throw an normal error and catch it in line 56 where you will send that error through response. Then we don't need to check for that condition again.
And all the blobKey related stuff will be in one method.